### ISSUE TITLE
`v2.0.0`-Release <- Read this for performance problems

### ISSUE THREAD

**Issue by CommanderStorm** [louislam/uptime-kuma#4500]
<sup>

This was originally a [PR here](https://github.com/louislam/uptime-kuma/pull/4171), but has been transformed into an issue due to being pin-able

</sup>

# what has been done?

We are hard at work improving Uptime Kuma.
If you want to have a [look at the full set of features for `v2.0`, have a look here](https://github.com/louislam/uptime-kuma/milestone/24).

Most notably this includes (**a full migration guide will be made before the release, dont worry**):
- better performance through [external MariaDB](https://github.com/louislam/uptime-kuma/pull/3017), [storing heartbeats in an aggregated form](https://github.com/louislam/uptime-kuma/pull/2750), [full server-side pagination for important events](https://github.com/louislam/uptime-kuma/pull/3515), [better async and database usage](https://github.com/louislam/uptime-kuma/pull/5025)
  => even though benchmarking is still out, we are confident that this pushes the prominent "limits" (**highly hardware-dependent, not a limit imposed by us**) ~500 Monitor or ~1.5GB DB-size
  A large part of the focus in this release is on performance. We don't know if we are optimising for the right thing, see #4456 for a discussion on this topic.
> [!TIP]
> If you are affected by the performance problems of `v1`, you need to reduce the amount of data you store.
The core problem is that Uptime Kuma in `v1` has to do a table scan of the entire heartbeat table for some operations.
>  
> The solution boils down to having a lower amount of data => making Uptime Kuma less worried about reading those gigabytes of data.:
> - **reduce the retention** and **execute a manual cleanup via the `VACUUM` button** under `/settings/monitor-history`
> - **Increase the time between checks**
> - **pause** or **reduce the amount of monitors**
> - **delete the specific history** of less essential monitors (to "lower" their retention below the configured maximum)

- tightening the security via offering rootless docker images images [#2086](https://github.com/louislam/uptime-kuma/pull/2086) [#4052](https://github.com/louislam/uptime-kuma/pull/4052)
- deprecated backup feature has been dropped (see https://github.com/louislam/uptime-kuma/pull/3892#issuecomment-1949941292 for an explanation why)
- deprecated cacheable lookups (not the same as `nscd`) have been dropped #3762
- A lot of bugfixes, some minor new features (see [the milestone](https://github.com/louislam/uptime-kuma/milestone/24))

# what is needed to get `2.0.0-beta.0` released

> [!NOTE]
> We hope `2.0.0-beta.0` can be released in spring 2024, which doesn't include the e2e tests yet. During the beta test, we will try to work on implementing the e2e tests.

> [!TIP]
> **How can I get involved?**
>
> While most of the bugs noted below require reading some code
> => gaining some context about how the aggregated tables work, some work do not and are such better starting points:
> - writing a synthetic benchmark which spawns some monitors and looks how the server handles it
> - helping in writing e2e tests using playwright

- some features have not adopted to new aggregate tables yet
  - [ ] Migrate existing heartbeat data into aggregate tables  
  - [x] Clear stats (#4324)
  - [ ] History retention 
  - [x] Ping chart (#4264)
  - [ ] ~Heartbeat table keeps 24-48 hours data only (excepted for important events, it will keep longer for the event list)~ (Can be deferred)
  - [x] Test and fix the Badges feature (originally introduced in #1119, a big headache for us right now because the new table schemas break this feature a lot) (https://github.com/louislam/uptime-kuma/pull/4850)
- [ ] Testing
  - [x] https://github.com/louislam/uptime-kuma/pull/2174
  - [x] https://github.com/louislam/uptime-kuma/pull/3893
  - [ ] Write more E2E Testing using Playwright
- [ ] Document how the aggregated tables work for our contributors in our [contribution guide](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md)
- [ ] Migration guide from `v1` to `v2` (Wiki) https://github.com/louislam/uptime-kuma-wiki/pull/53
  - [ ] New Docker tags
  - [ ] Notify about breaking changes and write a migration guide
    - removed variable badges
    - SMTP-templating #3414
      - dropped support for `vArIaBlE Cased` variables
      - deprecated `UPPERCASE` variables (this was added to make the `v1`⇒`v2` migration more seamless, but will be removed in v3)
    - removed deprecations:
      - #3762
      - #3892
    - https://github.com/louislam/uptime-kuma/pull/4289
    - default retires is now 0 instead of 1
- [ ] Benchmark where the new performance limits are
- [ ] Write the changelog and Release `2.0.0-beta.0`

**Comment by compgeniuses** [comment]
is version already installable at the moment, or there are many things yet to be done to make it so and for a pre-release

**Comment by CommanderStorm** [comment]
@compgeniuses 
Our installation guide only works for stable versions (it checks out a git tag to avoid users being on the wrong branch in `npm run setup`).
Pre-releases will be made once we don't have glaring bugs left, see my post above for which ones there are.

If you want to test-drive or contribute one of the areas named above, our [contribution guide](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) should have all the answers. If you are missing something for development, feel free to ping me.
We also have this way of testing PRs: https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests

**Comment by Aj7Ay** [comment]
Uptime kuma supports API call
so we can use Rundeck to set Incidents on it and once it resolves we can stop the incident 
is that possible?

**Comment by CommanderStorm** [comment]
This is an `discussion` issue. [Please ask `help`-quesstions via the appropirate means to not spam everyone who subscribes to this thread](https://github.com/louislam/uptime-kuma/issues/new?assignees=&labels=help&projects=&template=ask-for-help.yaml)

> Uptime kuma supports API call

Currently, we don't have an official API. Please refer to my answer to you here https://github.com/louislam/uptime-kuma/issues/118#issuecomment-1994429676

**Comment by VermiumSifell** [comment]
How is it going?

**Comment by CommanderStorm** [comment]
I think you are asking about a status update: The checklist above is kept up to date in what needs to happen before `2.0.0-beta.0`.

Is there something that we are missing in the post?

**Comment by VermiumSifell** [comment]
No I don't think so. But I am more interested in knowing if the beta is still estimated during the spring now?

**Comment by CommanderStorm** [comment]
We don't give out concrete deadlines and prioritise long-term team health over achieving short-term goals.
=> Maybe. See [#noestimates](https://www.youtube.com/watch?v=QVBlnCTu9Ms) why we don't give out more concrete timelines rather than "Spring" (=a flexgoal).

Being cheeky: If you want to move this along, you could consider contributing...

**Comment by timbo-tj** [comment]
What is the best way to try this out? We deploy via Docker on Railway.app and we use it to monitor our services. It is pretty low-stakes environment and I am happy to risk running a preview version of Uptime Kuma to get access to the new features and 'test run' it. But since it is not available on Docker hub I am not sure how to get started

**Comment by Zaid-maker** [comment]
> What is the best way to try this out? We deploy via Docker on Railway.app and we use it to monitor our services. It is pretty low-stakes environment and I am happy to risk running a preview version of Uptime Kuma to get access to the new features and 'test run' it. But since it is not available on Docker hub I am not sure how to get started

U can do this by not using docker but by running in nodejs environment. Read the README

**Comment by timbo-tj** [comment]
> > What is the best way to try this out? We deploy via Docker on Railway.app and we use it to monitor our services. It is pretty low-stakes environment and I am happy to risk running a preview version of Uptime Kuma to get access to the new features and 'test run' it. But since it is not available on Docker hub I am not sure how to get started
> 
> U can do this by not using docker but by running in nodejs environment. Read the README

Thanks - is it just the master branch? I couldn't find a specific branch or tag for v2.

**Comment by CommanderStorm** [comment]
In `master`, development happens for `v2.0-dev`.
We currently have NOT published a git/docker tag as we are NOT ready to have a public beta (see the checklist above)

You can:
- Test PRs via [the path listed here](https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests)
- help developing using the path laid out [in our contribution guide](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md)
- subscribe to [this issue](https://github.com/louislam/uptime-kuma/issues/4500#:~:text=a%20pull%20request.-,Notifications,-Customize) or to [releases](https://github.com/louislam/uptime-kuma) to get notified of the beta once it releases

> [!NOTE]
> If you want to use uptime kuma in any production environment, please refer to our [installation guide](https://github.com/louislam/uptime-kuma/wiki/%F0%9F%94%A7-How-to-Install).
> We do not offer stability/semver guarantees for development versions such as `v2.0-dev`.

**Comment by neelbhanushali** [comment]
@CommanderStorm 

I would like to take care of testing activity

Backend tests are stored in `test/backend-test` and Frontend tests are stored in `test/e2e` AFAIK

Can you help me with lists of tests that are to be written and documentation/examples of how they are to be written and tested

**Comment by CommanderStorm** [comment]
I am currently quite busy with uni stuff (I have a thesis deadline on Saturday) and I will get back to you on what is necessary precisely and on the code examples:

From the api side, I think these things need a test case:
- the badges API does work
- some of the other apis likely also benefit from testing, but would need to look into them 

I think the most beneficial would be a frontend end to end test of interacting with monitors:
- logging into the ui, adding a monitor
- seeing that the monitor does a ping
- enacting a maintenance period
- seeing that the monitor is in maintenance
- adding it to a status page

Fom the initialisation side (bit trickier):
- launching (fresh) via maridb works
- launching (fresh)
- migrating from one to the other works (don't think that we have yet looked into this)

From a monitor perspective (not related to v2.0, but to be complete):
Most of our monitors are currently untested. At least some of the bug reports are likely due to this.
My longer term plan is to use test containers to verify the monitors work as intended. I have currently not been able to finish the debugging effort of said PR (it is basically done, but one test case is not working. Will need to debug)

From the notification providers (not related to v2.0, but to be complete):
Our current testing strategy is to require screenshots of the relevant events by our contributors. I am unsure if we can test this better, but am open to ideas ^^

**Comment by CommanderStorm** [comment]
We don't give out concrete deadlines and prioritise long-term health over achieving short-term goals. In a volunteer run org, anything else would not be sustainable.
See [#noestimates](https://www.youtube.com/watch?v=QVBlnCTu9Ms) why we don't give out more concrete timelines rather than "Spring"/"Summer" (=a flexgoal).

Being cheeky: If you want to move this along, you could consider contributing ...

**Comment by gustavovalverde** [comment]
I'd like to know if it's possible to have v2 (or nightly docker versions) of this repo published in Docker Hub. If there's any technical constraint for that purpose, I might be able to contribute.

**Comment by compgeniuses** [comment]
> I'd like to know if it's possible to have v2 (or nightly docker versions) of this repo published in Docker Hub. If there's any technical constraint for that purpose, I might be able to contribute.

While a good idea, it may end up causing alot of support request that are unnecessary to the community due to incomplete build version.

**Comment by fabry006** [comment]
is there any ETA about v2.0.0 release date?
thanks

**Comment by apio-sys** [comment]
See top of this post, the tasklist progress is at 5/17 so still a lot of work todo. This is FOSS so to help speed up improvements, you can contribute in different areas as per your level of expertise. It has been explained several times that no ETA can be given depending on the contributions of the community mainly.

**Comment by CommanderStorm** [comment]
> tasklist progress is at 5/17 so still a lot of work todo

That is somewhat misleading.
There are two items left which NEED to happen:
- Going through the PRs, writing a changelog and migration guide
- Writing a db-migration for the existing heatbeat data into the new, aggregated schema.

literally any help is always appreciated.
Even if just looking at benchmarking/infra/testing, there are lots of bugs to catch (as with most FOSS software).

**Comment by FrancYescO** [comment]
> * **reduce the retention** and **execute a manual cleanup** under `/settings/monitor-history`

just to be clear, as i assume this is the best mitigation: after setting a lower number of retention days, for manual cleanup you intend a "clear full statistic" or "shrink database"? so the retention days are not applied to the previous stored datas?

**Comment by CommanderStorm** [comment]
Cleanup = the `VACUUM` button

**Comment by luke3butler** [comment]
> > See top of this post
> 
> Not even that, have commented somethng similar why we won't give out ETAs multiple times in this thread.. Example: [#4500 (comment)](https://github.com/louislam/uptime-kuma/issues/4500#issuecomment-2054135888)
> 
> > tasklist progress is at 5/17 so still a lot of work todo
> 
> That is somewhat misleading. There are two items left which NEED to happen:
> 
> * Going through the PRs, writing a changelog and migration guide
> * Writing a db-migration for the existing heatbeat data into the new, aggregated schema.
> 
> **literally any help is appreciated.** Even if just looking at benchmarking/infra/testing, there are lots of bugs to catch (as with most FOSS software).

I've written some code that fetched all the PRs since the latest tagged commit, outputs them into markdown files, adds code diffs, and then generates changelogs using Sonnet 3.5.

Here's some example outputs:

```
PR #4850
- Fixed uptime & ping badge duration to handle numeric input without units
- Implemented validation to ensure duration input is within allowed range
- Enabled support for different duration units (m, h, d, w, M, y)
- Updated badge label formatting to remove redundant unit suffix

PR #4757
- Bug fix: Corrected a typo in the internal code for handling the file upload icon for the status page.

PR #4728
- Improved MariaDB connection handling by configuring the connection pool with sensible defaults (min: 0, max: 10, idleTimeoutMillis: 30000)
- Addressed the `Error: read ECONNRESET` issue with Knex, which sometimes occurred when connecting to MariaDB
```

My plan was to then process these outputs again to combine and condense them into a single changelog. I think Sonnet might be capable of writing the database migration and migration guide as well.

1. Is this helpful?
2. If so, are there any special requests (formatting, adjustments, or otherwise) before I proceed?

**Comment by FrancYescO** [comment]
> Cleanup = the `VACUUM` button

Btw, this is a very temporary workaround... i have a 2,7GB (?! or at least is what is saying on the shrink button) database, my history was with no limit and i've set it to 90days (maybe there was a 2y or more history) but:
- the cleanup will likely lock all monitors for ~2h than all keep working good for just 2days than the dashboard will be again "unlodable" (monitors/notifications will keep working good)
- idk why the database is that big and did not decreased the size after setting a max days monitor history

**Comment by Ember-ruby** [comment]
> I've written some code that fetched all the PRs since the latest tagged commit, outputs them into markdown files, adds code diffs, and then generates changelogs using Sonnet 3.5.

> My plan was to then process these outputs again to combine and condense them into a single changelog. I think Sonnet might be capable of writing (or assisting to write) a migration guide as well.
> 
> Is this helpful?
> If so, are there any special requests (formatting, adjustments, or otherwise) before I proceed?

looking sonnet up, it seems to be an ai?
if you're using an ai for a changelog, check that it's actually correct -- which means *manually* going through each PR. 

but more importantly, a migration guide, which can have *serious* consequences if if's wrong, should *never* be written with an ai, or other automation.

**Comment by Wraaath** [comment]
> > I've written some code that fetched all the PRs since the latest tagged commit, outputs them into markdown files, adds code diffs, and then generates changelogs using Sonnet 3.5.
> 
> > My plan was to then process these outputs again to combine and condense them into a single changelog. I think Sonnet might be capable of writing (or assisting to write) a migration guide as well.
> > Is this helpful?
> > If so, are there any special requests (formatting, adjustments, or otherwise) before I proceed?
> 
> looking sonnet up, it seems to be an ai? if you're using an ai for a changelog, check that it's actually correct -- which means _manually_ going through each PR.
> 
> but more importantly, a migration guide, which can have _serious_ consequences if if's wrong, should _never_ be written with an ai, or other automation.

Whoever writes the changelog would have to go through every PR manually anyway. Let's just make sure together that the migration guide doesn't have any errors.

**Comment by pt-vamshi** [comment]
@CommanderStorm  Hello, has a release date been set for the V2 version? If not, when can we expect the V2 version to be available?

**Comment by zorrobiwan** [comment]
https://github.com/louislam/uptime-kuma/issues/4500#issuecomment-2054135888

https://github.com/louislam/uptime-kuma/issues/4500#issuecomment-2201448336

**Comment by zorrobiwan** [comment]
> @CommanderStorm Hello, has a release date been set for the V2 version? If not, when can we expect the V2 version to be available?

https://github.com/louislam/uptime-kuma/issues/4500#issuecomment-2054135888

https://github.com/louislam/uptime-kuma/issues/4500#issuecomment-2201448336

**Comment by CommanderStorm** [comment]
Closing as [`2.0.0-beta.0`](https://github.com/louislam/uptime-kuma/releases/tag/2.0.0-beta.0) was released

**Comment by otbutz** [comment]
Seems to embedded:

https://github.com/louislam/uptime-kuma/blob/4e144ec71ceb917fd08a1842a5f50538af8e5472/docker/debian-base.dockerfile#L69

I'd love if we could separate this.

**Comment by zenkraker** [comment]
personally, I'd prefer to either choose between using the embedded SQL-Lite (which we know the limitations in production, but good enough for a testing environment) OR go full power and leave the user running a separated MariaDB offering the following advantages (to my eyes of course):

1. UptimeKuma image will be smaller because you won't need to bloat it with mariaDB
2. Users can have control on the DB and backup is easier, because you just need to backup the volume of the MariaDB container and probably it is beneficial for uptimeKuma developers because, they won't need to think how to backup the DB or implement any feature for this and let the end-users deal with it

To summarize:
I'd add an env variable to uptimeKuma configuration, if it is filled it will use the mariaDB container + port, otherwise it will run with SQL-Lite by default

**Comment by CommanderStorm** [comment]
If you don't want mariadb, please use `beta-slim` instead.

> But in general, the peculiarity of docker is the detachment of dependencies so having an image ALL-IN-ONE is in my opinion against this principle

Having all of the features in the default image and having slim variants is quite common.

**Comment by zenkraker** [comment]
@CommanderStorm thanks for the response.
So, if I understood correctly, the slim version uses sql lite or can I pass set a connection to an external mariaDB container?

**Comment by CommanderStorm** [comment]
The slim version can use sqlite or an external mariaDB container

![image](https://github.com/user-attachments/assets/b4a1427e-1ec2-418c-ab86-5bd329976c54)

**Comment by zenkraker** [comment]
@CommanderStorm 
Thanks for the hint, the beta version works like a charm! <3

**Comment by msaleemCinnova** [comment]
@louislam , @CommanderStorm , 
I am currently running Uptime Kuma version 1.23.11 without Docker and would like to migrate to version 2. Could you kindly advise which branch I should switch to in order to access the beta/v2 release? Additionally, any guidance on the migration process would be greatly appreciated.

Thank you in advance for your support.

**Comment by CommanderStorm** [comment]
Please refer to https://github.com/louislam/uptime-kuma/wiki/Migration-From-v1-To-v2
If anything is missing, feedback is appreciated

**Comment by msaleemCinnova** [comment]
Dear Frank,

I am running Uptime Kuma in a non-Docker environment because I have integrated a custom module into the existing setup and submitted a pull request for these changes.[https://github.com/louislam/uptime-kuma/pull/5382]

To maintain my customizations, I update and manage the application without Docker. So i want to migrate from v1 to v2 without using Docker.

Here are the steps I follow for setup and execution & updates:

# Update from git
git fetch --all
git checkout 1.23.15 --force

# Install dependencies and prebuilt
npm install --production
npm run download-dist

npm run setup

# Option 1. Try it
node server/server.js
# Start Server
pm2 start server/server.js --name uptime-kuma

Looking for the migration guide without using Docker.




________________________________

**Comment by Wraaath** [comment]
> @CommanderStorm , Dear Frank,
> 
> I am running Uptime Kuma in a non-Docker environment because I have integrated a custom module into the existing setup and submitted a pull request for these changes.[https://github.com/[/pull/5382](https://github.com/louislam/uptime-kuma/pull/5382)]
> 
> To maintain my customizations, I update and manage the application without Docker. So i want to migrate from v1 to v2 without using Docker.
> 
> Here are the steps I follow for setup and execution & updates:
> # Update from git
> 
> git fetch --all git checkout 1.23.15 --force
> # Install dependencies and prebuilt
> 
> npm install --production npm run download-dist
> 
> npm run setup
> # Option 1. Try it
> 
> node server/server.js
> # Start Server
> 
> pm2 start server/server.js --name uptime-kuma
> 
> Looking for the migration guide without using Docker.

You could build your own Docker image with the patch maybe? Unless you really don't wanna use Docker

**Comment by CommanderStorm** [comment]
It sounds like you are looking for a development environment.
We have documented this here: https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md

What you are trying to do is otherwise not well supported as it is not the main path.
You can extend it, but make sure you actually have the state as you expect it. Both `npm run setup` and the git part may reset your changes.
In any case, there is no official support from us for this kind of deployment as I don't have capacity for this.

**Comment by msaleemCinnova** [comment]
@CommanderStorm ,
Thank you for your response and guidance.

I wanted to let you know that I have already opened a pull request (PR) to contribute to the project. My goal is to enhance Uptime Kuma by adding SIP-related features, and I fully understand that my deployment method is not the standard approach.

However, I'm currently running Uptime Kuma using PM2 in a Linux environment, not Docker, and I aim to migrate from v1 to v2 without relying on Docker. I realize this isn't officially supported

**Comment by msaleemCinnova** [comment]
@Wraaath , 
I’m currently running Uptime Kuma using PM2 in a Linux environment instead of Docker, and my goal is to migrate from v1 to v2 while continuing to use PM2. I prefer not to rely on a custom Docker image with patches for long-term use. However, if creating a custom image is necessary to complete the migration process, I’m open to exploring that option temporarily.

Could you please clarify what specific patch or adjustments you are referring to? Your advice will help me ensure a smooth migration while aligning with the project’s standards.

**Comment by CommanderStorm** [comment]
@msaleemCinnova 
Sorry, did not have the time to look into said PR. I am currently quite overloaded with Uni and other work.

Since you already wrote the code, please just post it. Having a discussion is not quite relevant to adding notification providers or basic (cannot be implemented using current monitors) monitors like the one you are trying to add.

I don't know what a SIP is => please preferably include a test case that it works (see the mqtt test cases for examples how this can be done).

**Comment by msaleemCinnova** [comment]
@CommanderStorm ,
Thank you for taking the time to respond despite your busy schedule—I completely understand how demanding university and other work commitments can be.

I will add the code to submitted the PR. I’ll respect your preference to keep the process straightforward and will avoid extensive discussions.

**Comment by IAmBecomeDeth** [comment]
> [@msaleemCinnova](https://github.com/msaleemCinnova) Sorry, did not have the time to look into said PR. I am currently quite overloaded with Uni and other work.
> 
> Since you already wrote the code, please just post it. Having a discussion is not quite relevant to adding notification providers or basic (cannot be implemented using current monitors) monitors like the one you are trying to add.
> 
> I don't know what a SIP is => please preferably include a test case that it works (see the mqtt test cases for examples how this can be done).

I am following the pull request mentioned above, and would GREATLY appreciate if it were tested then merged into the codebase. 

SIP is for VOIP phones/servers. most VOIP phones (not all) use sip. Looking at the screenshots in the PR it looks functional complete but I have not tried running it myself. 

I use uptimekuma to monitor the websites on my web server and the web interface on my voip server but that doesn't necessarily tell me whether people can connect to the server for phone calls. I would love to use Uptime Kuma to monitor the sip availability as well!

**Comment by CommanderStorm** [comment]
@IAmBecomeDeth you can test PRs via the instructions found here: https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests
